### DISCUSSION TITLE
Version 0.11.0-M1 causing script compilation error (BoundDep)

### DISCUSSION THREAD

**Discussion by hmf** [com-lihaoyi/mill#2213] [Q&A] [answered]
I have the following code snippet in my build file:

```scala
  def mdocSources = T.sources { super.millSourcePath / "docs" }
  def mdocDep: Task[Agg[Dep]] = T.task{ 
      Agg(
        ivy"org.scalameta::mdoc:${scalaMdocVersion}"
          .exclude("org.scala-lang" -> "scala3-compiler_3")
          .exclude("org.scala-lang" -> "scala3-library_3"),
        ivy"org.scala-lang::scala3-compiler:${scalaVersion()}"
      )
    } 

  def mDocLibs = T{ resolveDeps(mdocDep) }
```

which was working with version `10.10.0`, but now I get the error:

```
/home/hmf/VSCodeProjects/sploty/build.sc:145: type mismatch;
 found   : mill.define.Task[mill.Agg[mill.scalalib.Dep]]
    (which expands to)  mill.define.Task[mill.api.Loose.Agg[mill.scalalib.Dep]]
 required: mill.define.Task[mill.Agg[mil.scalalib.BoundDep]]
    (which expands to)  mill.define.Task[mill.api.Loose.Agg[mil.scalalib.BoundDep]]
  def mDocLibs = T{ resolveDeps(mdocDep) }
                                ^
/home/hmf/VSCodeProjects/sploty/build.sc:145: `T.command` definitions must have 1 parameter list
  def mDocLibs = T{ resolveDeps(mdocDep) }
      ^
Compilation Failed
```

So I changed `mdocDep` to: 

```scala
  def mdocDep: Task[Agg[mil.scalalib.BoundDep]] = T.task{ 
      Agg(
        ivy"org.scalameta::mdoc:${scalaMdocVersion}"
          .exclude("org.scala-lang" -> "scala3-compiler_3")
          .exclude("org.scala-lang" -> "scala3-library_3"),
        ivy"org.scala-lang::scala3-compiler:${scalaVersion()}"
      )
    } 
```

So now I get the error:

```
/home/hmf/VSCodeProjects/sploty/build.sc:136: type mismatch;
 found   : mill.api.Loose.Agg[mill.scalalib.Dep]
 required: mill.api.Result[mill.Agg[mil.scalalib.BoundDep]]
    (which expands to)  mill.api.Result[mill.api.Loose.Agg[mil.scalalib.BoundDep]]
      Agg(
         ^
Compilation Failed
```

Removing the explicit typing also did not help. A search in the documentation does not show anything. In the github repository, I did not find usage examples. The documentation does not seem to be changed in relation to setting up our libraries. 

What must I change to correct this?

TIA

**Comment by lefou** [comment, marked as answer]
It looks like you are using the in-development `main` tree, which is currently released as `0.11.0-M1` 

The API of `CoursierModule`, `JavaModule` and its derivations changed slightly, so you need to adapt to the new type signatures, if you want to keep up with the latest development. See PR #1574 for details. Unfortunately, I made a typo in the package name of `BoundDep`, which I later fixed in #2214.

There is currently no migration guide written, as Mill `0.11` isn't released yet.

(Please note that I'm going to delete your other discussion entry (#2212), as it is almost identical to this one.)

For a rough overview of the planned changes of Mill 0.11, you can also read the following discussion:

* #1826

**Reply by hmf** [reply]
@lefou Thanks for the information. 

> It looks like you are using the in-development main tree, which is currently released as 0.11.0-M1

Correct. Apologies for that, I forgot to add that to description. Just want to be prepared for the next release. 

> There is currently no migration guide written, as Mill 0.11 isn't released yet.

Ok. I will read up on #1574 and #1826. This last one I have read when updated, but seemed over my head. 

> (Please note that I'm going to delete your other discussion entry (#2212), as it is almost identical to this one.)

Oops. My mistake, it is a duplicate. Sorry about that. 

Thank you.
